src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+        return std::make_shared<DataTypeUInt64>();
+    else if (type == "<f2")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f4")
+        return std::make_shared<DataTypeFloat32>();
+    else if (type == "<f8")
+        return std::make_shared<DataTypeFloat64>();
+    else if (type == "<c8" || type == "<c16")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support complex numeric type");
+    else if (type == "|b1")
+        return std::make_shared<DataTypeInt8>();
+    else if (type[1] == 'U' || type[1] == 'S')
+        return std::make_shared<DataTypeString>();
+    else if (type == "O")
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "ClickHouse doesn't support object types");
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Error while parsing data type");
+}
+
+std::vector<int> parseShape(String shapeString)
+{
+    shapeString.erase(std::remove(shapeString.begin(), shapeString.end(), '('), shapeString.end());
+    shapeString.erase(std::remove(shapeString.begin(), shapeString.end(), ')'), shapeString.end());
+
+    // Use a string stream to extract integers
+    String value;
+
+    std::vector<int> shape;
+
+    size_t start = 0, end = 0;
+
+    while ((end = shapeString.find(',', start)) != std::string::npos)
+    {
+        shape.push_back(std::stoi(shapeString.substr(start, end - start)));
+        start = end + 1;
+    }
+
+    // Add the last token (or the only token if no delimiter is found)
+    if (start != shapeString.length())
+        shape.push_back(std::stoi(shapeString.substr(start)));
+    return shape;
+}
+
+std::unordered_map<String, String> parseHeader(ReadBuffer &buf)
+{
+    /// Check magic bytes
+    const char * magic_string = "\x93NUMPY";
+    assertString(magic_string, buf);
+
+    /// Read npy version.
+    UInt8 version_major;
+    UInt8 version_minor;
+    readBinary(version_major, buf);
+    readBinary(version_minor, buf);
+
+    /// Read header length.
+    UInt32 header_length;
+    /// In v1 header length is 2 bytes, in v2 - 4 bytes.
+    if (version_major == 1)
+    {
+        UInt16 header_length_u16;
+        readBinaryLittleEndian(header_length_u16, buf);
+        header_length = header_length_u16;
+    }
+    else
+    {
+        readBinaryLittleEndian(header_length, buf);
+    }
+
+    /// Remember current count of read bytes to skip remaining
+    /// bytes in header when we find all required fields.
+    size_t header_start = buf.count();
+
+    /// Start parsing header.
+    String shape;
+    String descr;
+
+    assertChar('{', buf);
+    skipWhitespaceIfAny(buf);
+    bool first = true;
+    while (!checkChar('}', buf))
+    {
+        /// Skip delimiter between key-value pairs.
+        if (!first)
+        {
+            skipWhitespaceIfAny(buf);
+        }
+        else
+        {
+            first = false;
+        }
+
+        /// Read map key.
+        String key;
+        readQuotedField(key, buf);
```

Comment:
Maybe for keys we can use `readQuotedString` instead of `readQuotedField `, because we know that all keys are strings (If it's not like this, correct me please), so `key` will not contain single quotes